Fix error caused by updating immutable object

Dexagod opened the pull request:
I fixed the updates to the immutable `Quad` object in the `Datasource.js` file by creating a new `Quad` object with the updated value.
I updated the tests for the `CompositeDatasource` to include a `dataFactory` field for the created datasources.
This closes #137

RubenVerborgh commented on packages/core/lib/datasources/Datasource.js:
```diff
@@ -137,18 +137,25 @@ class Datasource extends EventEmitter {
+      else if (graph && (!quad.graph || quad.graph.termType === 'DefaultGraph'))
+        transformedGraph = graph;
+
+      const transformedQuad = this.dataFactory.quad(
+        transformedSubject || quad.subject,
+        quad.predicate,
+        transformedObject || quad.object,
+        transformedGraph || quad.graph,
+      );
+      return transformedQuad;
```
```suggestion
      return this.dataFactory.quad(
        transformedSubject || quad.subject,
        quad.predicate,
        transformedObject || quad.object,
        transformedGraph || quad.graph,
      );
```

Dexagod replied:
Is this an indentation issue?

rubensworks replied:
No need to have `transformedQuad`.

Dexagod replied:
:sweat_smile:

RubenVerborgh commented on packages/core/lib/datasources/Datasource.js:
```diff
@@ -137,18 +137,25 @@ class Datasource extends EventEmitter {
     // Transform the received quads
     let destination = new BufferedIterator(), outputQuads, graph = this._graph;
     outputQuads = destination.map((quad) => {
+      let transformedSubject, transformedObject, transformedGraph;
```
Alternative way of writing the method could be to

```suggestion
      let { subject, predicate, object, graph } = quad;
```

so you don't have to use `quad.` all the time below, and the `\\` from the `transformedQuad` can also be dropped.

Dexagod replied:
It is indeed not the most elegant solution, but a graph variable was already declared for that function, and I wanted to make sure not to introduce inconsistencies or make the code difficult to understand. 
I could update it, and rename the previously introduced graph variable (that one refers to the default graph that has been set if I recall correctly)

Dexagod replied:
I changed the custom default graph variable to _defaultGraph_ instead of _graph_, and used this approach in latest commit

RubenVerborgh reviewed (approved):
Good work!

rubensworks commented on packages/core/lib/datasources/Datasource.js:
```diff
@@ -137,18 +137,25 @@ class Datasource extends EventEmitter {
+      let transformedSubject, transformedObject, transformedGraph;
       // Translate blank nodes in the result to blank node IRIs.
       if (quad.subject && quad.subject.termType === 'BlankNode' && !this._skolemizeBlacklist[quad.subject.value])
         quad.subject = this.dataFactory.namedNode(blankNodePrefix + quad.subject.value);
```
This will still produce an error.

Dexagod replied:
Updated the subject and object assignments too.
Any other way this produces an error

rubensworks replied:
If it runs and the tests pass with the latest N3 version, then it should be ok I guess.

rubensworks commented on packages/core/lib/datasources/Datasource.js:
```diff
@@ -137,18 +137,25 @@ class Datasource extends EventEmitter {
       if (quad.subject && quad.subject.termType === 'BlankNode' && !this._skolemizeBlacklist[quad.subject.value])
         quad.subject = this.dataFactory.namedNode(blankNodePrefix + quad.subject.value);
       if (quad.object  && quad.object.termType  === 'BlankNode' && !this._skolemizeBlacklist[quad.object.value])
         quad.object  = this.dataFactory.namedNode(blankNodePrefix + quad.object.value);
```
Same as above.

rubensworks reviewed (changes requested):
You may also want to consider updating the lib (I assume N3.js) that provides the data factory.
Then you can effectively confirm that #137 is fixed.

rubensworks commented on packages/core/lib/views/RdfView.js:
```diff
@@ -86,8 +86,9 @@ class RdfView extends View {
-        quad.graph = quad.graph.termType === 'DefaultGraph' ? (metadataGraph ? dataFactory.namedNode(metadataGraph) : dataFactory.defaultGraph()) : quad.graph;
-        writer.addQuad(quad);
+        const graph = quad.graph.termType === 'DefaultGraph' ? (metadataGraph ? dataFactory.namedNode(metadataGraph) : dataFactory.defaultGraph()) : quad.graph;
+        const newQuad = dataFactory.quad(quad.subject, quad.predicate, quad.object, graph);
```
No need to have `newQuad`. (also below)

Dexagod commented:
I have addressed all issues and updated the N3 library and the relevant tests, but it seems that the last commit reduced the test coverage with .04%, not sure why

rubensworks commented:
Thanks, great work!
